src/bin/pg_autoctl/system_utils.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * src/bin/pg_autoctl/hardware_utils.c
+ *   Utility functions for getting CPU and Memory information.
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the PostgreSQL License.
+ *
+ */
+
+#if defined(__linux__)
+#include <sys/sysinfo.h>
+#else
+#include <sys/types.h>
+#include <sys/sysctl.h>
+#endif
+
+#include <math.h>
+
+#include "log.h"
+#include "file_utils.h"
+#include "system_utils.h"
+
+#if defined(__linux__)
+static bool get_system_info_linux(SystemInfo *sysInfo);
+#endif
+
+#if defined(__APPLE__) || defined(BSD)
+static bool get_system_info_bsd(SystemInfo *sysInfo);
+#endif
+
+/*
+ * get_system_info probes for system information and fills the given SystemInfo
+ * structure with what we found: number of CPUs and total amount of memory.
+ */
+bool
+get_system_info(SystemInfo *sysInfo)
+{
+#if defined(__APPLE__) || defined(BSD)
+	return get_system_info_bsd(sysInfo);
+#elif defined(__linux__)
+	return get_system_info_linux(sysInfo);
+#else
+	log_error("Failed to get system information: "
+			  "Operating System not supported");
+	return false;
+#endif
+}
+
+
+/*
+ * On Linux, use sysinfo(2) and getnprocs(3)
+ */
+#if defined(__linux__)
+static bool
+get_system_info_linux(SystemInfo *sysInfo)
+{
+	struct sysinfo linuxSysInfo = { 0 };
+
+	if (sysinfo(&linuxSysInfo) != 0)
+	{
+		log_error("Failed to call sysinfo(): %m");
+		return false;
+	}
+
+	sysInfo->ncpu = get_nprocs();
+	sysInfo->totalram = linuxSysInfo.totalram;
+
+	return true;
+}
+
+
+#endif
+
+
+/*
+ * FreeBSD, OpenBSD, and darwin use the sysctlbyname(3) API.
+ */
+#if defined(__APPLE__) || defined(BSD)
+static bool
+get_system_info_bsd(SystemInfo *sysInfo)
+{
+	unsigned int ncpu = 0;      /* the API requires an integer here */
+
+	size_t cpuSize = sizeof(ncpu);
+	size_t memSize = sizeof(sysInfo->totalram);
+
+	if (sysctlbyname("hw.ncpu", &ncpu, &cpuSize, NULL, 0) != 0)
+	{
+		log_error("Failed to probe number of CPUs: %m");
+		return false;
+	}
+
+	sysInfo->ncpu = (unsigned short) ncpu;
+
+	if (sysctlbyname("hw.memsize", &(sysInfo->totalram), &memSize, NULL, 0) != 0)
+	{
+		log_error("Failed to probe Physical Memory: %m");
+		return false;
+	}
+
+	return true;
+}
+
+
+#endif
+
+
+/*
+ * pretty_print_bytes pretty prints bytes in a human readable form. Given
+ * 17179869184 it places the string "16 GB" in the given buffer.
+ */
+void
+pretty_print_bytes(char *buffer, size_t size, uint64_t bytes)
```

Comment:
This rounds down `1.9999 GB` as a byte number to `1 GB`. Is that really what we want here?

Reply:
I just pushed a change that should now display `'1999 MB'`, and display `4096 MB` where it would before display `4 GB`. Please try the changes?

Reply:
Cool on my machine I now get `1989 MB` shared_buffers

Reply:
I'm now thinking that we might need a way to instruct pg_autoctl to not do the tuning... can you still create a monitor and 3 postgres or 4 nodes locally for testing? ;-)